frame/collective/src/lib.rs

```diff
@@ -182,72 +306,200 @@ decl_module! {
+		) -> DispatchResultWithPostInfo {
 			let who = ensure_signed(origin)?;
-			ensure!(Self::is_member(&who), Error::<T, I>::NotMember);
+			let members = Self::members();
+			ensure!(members.contains(&who), Error::<T, I>::NotMember);
 
 			let proposal_hash = T::Hashing::hash_of(&proposal);
-
 			ensure!(!<ProposalOf<T, I>>::contains_key(proposal_hash), Error::<T, I>::DuplicateProposal);
 
 			if threshold < 2 {
 				let seats = Self::members().len() as MemberCount;
-				let ok = proposal.dispatch(RawOrigin::Members(1, seats).into()).is_ok();
-				Self::deposit_event(RawEvent::Executed(proposal_hash, ok));
+				let result = proposal.dispatch(RawOrigin::Members(1, seats).into());
+				Self::deposit_event(RawEvent::Executed(proposal_hash, result.is_ok()));
+
+				Ok(get_result_weight(result).map(|w| weight_for::propose_execute(
+					T::DbWeight::get(),
+					members.len() as Weight, // M
+					w, // P1
+				)).into())
 			} else {
+				let active_proposals =
+					<Proposals<T, I>>::try_mutate(|proposals| -> Result<usize, DispatchError> {
+						proposals.push(proposal_hash);
+						ensure!(
+							proposals.len() <= T::MaxProposals::get() as usize,
+							Error::<T, I>::TooManyProposals
+						);
+						Ok(proposals.len())
+					})?;
 				let index = Self::proposal_count();
 				<ProposalCount<I>>::mutate(|i| *i += 1);
-				<Proposals<T, I>>::mutate(|proposals| proposals.push(proposal_hash));
 				<ProposalOf<T, I>>::insert(proposal_hash, *proposal);
 				let end = system::Module::<T>::block_number() + T::MotionDuration::get();
 				let votes = Votes { index, threshold, ayes: vec![who.clone()], nays: vec![], end };
 				<Voting<T, I>>::insert(proposal_hash, votes);
 
 				Self::deposit_event(RawEvent::Proposed(who, index, proposal_hash, threshold));
+
+				Ok(Some(weight_for::propose_proposed(
+					T::DbWeight::get(),
+					members.len() as Weight, // M
+					active_proposals as Weight // P2
+				)).into())
 			}
 		}
 
+		/// Add an aye or nay vote for the sender to the given proposal.
+		///
+		/// Requires the sender to be a member.
+		///
 		/// # <weight>
-		/// - Bounded storage read and writes.
-		/// - Will be slightly heavier if the proposal is approved / disapproved after the vote.
+		/// ## Weight
+		/// - `O(M)` where `M` is members-count (code- and governance-bounded)
+		/// - DB:
+		///   - 1 storage read `Members` (codec `O(M)`)
+		///   - 1 storage mutation `Voting` (codec `O(M)`)
+		/// - 1 event
+		/// - Benchmark: 24.03 + M * 0.349 + P * 0.119 + B * 0.003 µs (min squares analysis)
 		/// # </weight>
-		#[weight = (200_000_000, DispatchClass::Operational)]
-		fn vote(origin, proposal: T::Hash, #[compact] index: ProposalIndex, approve: bool) {
+		#[weight = (
+			weight_for::vote(T::DbWeight::get(), T::MaxMembers::get()),
+			DispatchClass::Operational
+		)]
+		fn vote(
+			origin, proposal: T::Hash, #[compact] index: ProposalIndex, approve: bool
```

Comment:
I believe it is normal "styling" to do the following:

* Keep origin with the fn, since it is not like. a normal parameter and more like macro magic
* give each parameter its own line if you need a newline to begin with